### ISSUE TITLE
Data exported with dump datasets should be importable with load datasets

### ISSUE THREAD

**Issue by sirex** [ckan/ckanapi#106]
Documentation says:

> Example: dumping datasets from CKAN into a local file with 4 processes:
>
>     $ ckanapi dump datasets --all -O datasets.jsonl.gz -z -p 4 -r http://localhost
>
> Example: load datasets from a dataset dump file with 3 processes in parallel:
>
>     $ ckanapi load datasets -I datasets.jsonl.gz -z -p 3 -c /etc/ckan/production.ini

But unfortunately what was exported with `dump datasets` can't be imported to an empty CKAN database with `load datasets`. There are [some workarounds](https://github.com/ckan/ckanapi/issues/52#issuecomment-125027007), but I think importing data that was exported using same tool should just work without any workarounds.

**Comment by wardi** [comment]
I use ckanapi exactly the way it's documented and it works for me because I've also imported the same organizations with the same ids.

If you'd like to use the organization names instead of their ids you'll have to modify the output, but I don't want to make the ckanapi tool do that automatically because that would be changing the data that CKAN itself is returning. If you think CKAN should be returning different data, that's a ckan issue, not a ckanapi one.

**Comment by wardi** [comment]
We could change load datasets slightly so that it tries again with the org name replacing the owner_org field when getting this specific error, though. I'd be ok with that.

**Comment by sirex** [comment]
Here is what I tried to do:

1. Exported data from an existing CKAN instance:
        
        ckanapi dump datasets --all -O data/opendata.jsonl -r http://opendata.lt/

2. Installed new CKAN locally and tried to import previously exported data:

        ckanapi load datasets -I data/both.jsonl -c development.ini

But nothing got imported:

    1 [2] --- create ValidationError {"owner_org":["Organization does not exist"]}
    2 [3] 0.03s create ValidationError {"owner_org":["An organization must be provided"]}
    3 [4] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    4 [5] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    5 [6] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    6 [7] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    7 [8] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    8 [9] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    9 [10] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    10 [11] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    11 [12] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    12 [13] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    13 [14] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    14 [15] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    15 [16] 0.04s create ValidationError {"owner_org":["Organization does not exist"]}
    16 [17] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    17 [18] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    18 [19] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    19 [20] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    20 [21] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    21 [22] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    22 [23] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    23 [24] 0.04s create ValidationError {"owner_org":["Organization does not exist"]}
    24 [25] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    25 [26] 0.03s create ValidationError {"owner_org":["Organization does not exist"]}
    26 [None] 0.02s create ValidationError {"owner_org":["An organization must be provided"]}

**Comment by wardi** [comment]
@sirex right, so you're missing the organizations that exist on http://opendata.lt/ (the Org does not exist errors) and you're requiring orgs in your local ckan (the org must be provided errors)

**Comment by sirex** [comment]
@wardi yes, I'm importing data into completely empty CKAN instance. I see that exported dump has all the information needed to create organizations.

In general I expect this to work like `pg_dump > dump.sql`, `psql < dump.sql`.

**Comment by wardi** [comment]
Your datasets dump is just datasets. It's not your full database. The organization details in the dataset records are truncated and include only a few fields, like title and description, just for convenience.

Try `ckanapi dump organizations --all ... | ckanapi load organizations` first if you want to be able to load datasets that belong to organizations.

**Comment by sirex** [comment]
Aha, thanks, finally I got it... :)

So basically what I need is `ckanapi dump everything | ckanapi load everything`, instead of `ckanapi dump (datasets | groups | organizations | users | related) ...`.

I think this would be nice to have a command that dumps and imports everything, so that you could easily transfer data between CKAN instances. As I understand this is not implemented.

**Comment by wardi** [comment]
Not as such, but if you're going to do it I'd suggest doing it in this order:
1. users
2. organizations (users' memberships will be assigned)
3. datasets (organizations are required)

related is an old type that you'll likely not need. If you're using ckanext-showcase your data will already be covered by the dataset dump.

**Comment by sirex** [comment]
Ok, then I will try to implement `ckan dump` for dumping everything and `ckan load` for loading everything what was dumped. It is quite possible that I will give up if it takes too long, but I will try.

**Comment by NicolaiMogensen** [comment]
Is something like this implemented yet? I still think it would be a nice feature to easily dump and load ckan instances.

**Comment by torfsen** [comment]
@NicolaiMogensen If you really want to dump all of CKAN's data then you're better off with using [PostgreSQL's import/export tools](http://docs.ckan.org/en/latest/maintaining/database-management.html#import-and-export) directly. `ckanapi` helps you to export and import specific parts of that data (e.g. datasets), but support for each of these is added manually to `ckanapi`. For example, a lot of CKAN extensions add their own tables to CKAN's database, something that `ckanapi` doesn't take into consideration (by design).

**Comment by NicolaiMogensen** [comment]
@torfsen Okay! This is actually what i have been doing so far. What about the resources and storage folders? Is it plug-and-play in the sense that i can just copy those folders from one instance to the other? I have been trying different methods of copying using this api, scripting etc to find an optimal solution.

**Comment by torfsen** [comment]
@NicolaiMogensen My usual content transfer workflow is to replicate the database via the mentioned PostgreSQL tools and to copy the storage directory. Afterwards I rebuild the Solr index. That works fine for backups and similar tasks that are not time critical. Real-time replication, e.g. for a hot failover, would obviously need a different strategy.
